### PR TITLE
Add quads to vtk writer

### PR DESCRIPTION
**Pull request by Davknapp**
Add Quads to the vtk-writer. Similar to the triangle-vtk-writer.

### CODE REVIEW

**Comment by jlchan** [comment]
@Davknapp it looks like this has some conflicts, and seems to duplicate some of the changes made in #81 too. Is there any way to fix this? For example, could you branch off of `main` now that #81 is merged and cherry pick the new commits from this PR?

**Comment by Davknapp** [comment]
> @Davknapp it looks like this has some conflicts, and seems to duplicate some of the changes made in #81 too. Is there any way to fix this? For example, could you branch off of `main` now that #81 is merged and cherry pick the new commits from this PR?

I merged the current main-branch into this branch. The branch is directly build on #81 therefore it should solve the issue.

File: src/mesh/vtk_helper.jl [hunk 1]
```diff
@@ -80,6 +126,17 @@ function vtk_order(elem::Tri, order)
     return triangle_vtk_order(tri_sud_vertices, order, 2)
 end
 
+"""
+    vtk_order(elem::Quad, order)
+
+Construct all node-points of a VTK_LAGRANGE_TRIANGLE of order order. The corner-nodes are
+given by the reference-triangle used by StartUpDG
```
Comment (jlchan) [inline src/mesh/vtk_helper.jl:133]:
```suggestion
Construct all node-points of a VTK_LAGRANGE_QUAD of order `order`. The corner-nodes are
given by the reference quadrilateral used by StartUpDG
```

File: test/write_vtk_tests.jl [hunk 2]
```diff
@@ -2,34 +2,48 @@ function quad(x, y)
...
-        elseif order == 0
-            @test triangle_vtk_order(tri_sud_vertices, order, 2) == [-1.0; -1.0]
-        else
-            @test triangle_vtk_order(tri_sud_vertices, order, 2) == tri_sud_vertices
+@testset "VTKWriter test for $elem" for elem in [Tri(), Quad()]
+    @testset "Polygrad $N" for N = [1, 2, 3, 4]
+        @testset "Write Mesh" begin
+            rd = RefElemData(elem, N)
+            md = MeshData(uniform_mesh(elem, 4)..., rd)
+            interpolate = vandermonde(rd.element_type, rd.N, equi_nodes(rd.element_type, rd.N)...) /rd.VDM
```
Comment (jlchan) [inline test/write_vtk_tests.jl:28]:
```suggestion
            interpolate = vandermonde(rd.element_type, rd.N, equi_nodes(rd.element_type, rd.N)...) / rd.VDM
```

File: src/mesh/vtk_helper.jl [hunk 3]
```diff
@@ -69,6 +69,52 @@ function triangle_vtk_order(corner_verts, order, dim, skip = false)
...
+function quad_vtk_order(corner_verts, order, dim, skip = false)
+    coords = Matrix{Float64}(undef, 0, dim)
+    if order < 0
+        return nothing
+    end
+    if skip == false
+        coords = copy(corner_verts)
+    end
+    num_verts_on_edge = order - 1
+    edges = [(1,2), (2,3), (4,3), (1,4)]
```
Comment (jlchan) [inline src/mesh/vtk_helper.jl:91]: This ordering is different from `StartUpDG.face_vertices(Quad())`. Does that matter? Is this just the ordering VTK uses?

**Review by jlchan — CHANGES_REQUESTED** [review]
This looks good! I only have a few questions about assumptions in node and face orderings.

File: src/mesh/vtk_helper.jl (hunk 3 above)
Comment (Davknapp) [reply src/mesh/vtk_helper.jl:91]: This is the vtk-ordering

File: src/mesh/vtk_helper.jl [hunk 4]
```diff
@@ -80,6 +126,17 @@ function vtk_order(elem::Tri, order)
...
 end
 
+"""
+    vtk_order(elem::Quad, order)
+
+Construct all node-points of a VTK_LAGRANGE_TRIANGLE of order order. The corner-nodes are
+given by the reference-triangle used by StartUpDG
+"""
+function vtk_order(elem::Quad, order)
+    quad_sud_vertices = [-1.0 1.0 1.0 -1.0; -1.0 -1.0 1.0 1.0]
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/mesh/vtk_helper.jl]: I think that my choice of variable-name and documentation wasn't perfect in this case. I updated it in the recent push and hope that it is better and more self-explainable now.

File: test/write_vtk_tests.jl [hunk 5]
```diff
@@ -2,34 +2,48 @@ function quad(x, y)
...
-        check = filename * ".vtu"
-        #Todo: Can we implement a better check?
-        vtu_name = MeshData_to_vtk(md, rd, pdata, ["(x+y)^2"], filename, true)
-        @test vtu_name[1] == check
+function deg_one_order(elem)
+    #If the order of the tri-vertices changes in StartUpDG update this line.
+    if elem == Tri()
+        return permutedims(hcat(nodes(Tri(), 1)...))
+    elseif elem == Quad()
+        return [-1.0 1.0 1.0 -1.0; -1.0 -1.0 1.0 1.0]
```
Comment (jlchan) [inline test/write_vtk_tests.jl:10]:
```suggestion
# test expected output of `vtk_order`
function deg_one_order(elem)
    if elem == Tri()
        return permutedims(hcat(nodes(Tri(), 1)...))
    elseif elem == Quad()
        return [-1.0 1.0 1.0 -1.0; -1.0 -1.0 1.0 1.0]
```

File: test/write_vtk_tests.jl [hunk 6]
```diff
@@ -2,34 +2,48 @@ function quad(x, y)
...
 
-@testset "Type to vtk" begin
-    @test type_to_vtk(Tri()) == VTKCellTypes.VTK_LAGRANGE_TRIANGLE
+function deg_zero_order(elem)
+    #If the order of the tri-vertices changes in StartUpDG update this line.
+    if elem == Tri()
+        return [-1.0; -1.0]
+    elseif elem == Quad()
+        return [-1.0 1.0 1.0 -1.0; -1.0 -1.0 1.0 1.0]
+    end
```
Comment (jlchan) [inline test/write_vtk_tests.jl:20]:
```suggestion
# test expected output of `vtk_order`
function deg_zero_order(elem)
    #If the order of the tri-vertices changes in StartUpDG update this line.
    if elem == Tri()
        return [-1.0; -1.0]
    elseif elem == Quad()
        return [-1.0 1.0 1.0 -1.0; -1.0 -1.0 1.0 1.0]
    end
```

**Review by jlchan — APPROVED** [review]
LGTM! @Davknapp I'll merge it in about 30 minutes unless you say otherwise

I'll also release a new version since this should cover 2D VTK visualization.

**Pull request merged**